### ISSUE TITLE
CL_DEVICE_TYPE_ALL and custom devices

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#300]
While working on #299 I found that querying `CL_DEVICE_TYPE_ALL` devices has this special case (**emphasis** mine):

> `CL_​DEVICE_​TYPE_​ALL`: All OpenCL devices available in the platform, **except for `CL_DEVICE_TYPE_CUSTOM` devices.**

Note that in the headers we have:

```c
/* cl_device_type - bitfield */
// snip
#ifdef CL_VERSION_1_2
#define CL_DEVICE_TYPE_CUSTOM                       (1 << 4)
#endif
#define CL_DEVICE_TYPE_ALL                          0xFFFFFFFF
```

This makes handling of `CL_DEVICE_TYPE_ALL` and `CL_DEVICE_TYPE_CUSTOM` awkward both for implementers and users - implementers cannot purely look at the bits of the passed-in _device_type_ to determine what is being queried, and users need to make two calls to `clGetDeviceIDs` to truly query for all of the devices in the platform.

Would it make more sense to eliminate the special handling for `CL_DEVICE_TYPE_CUSTOM`, and return **all** devices in the platform for `CL_DEVICE_TYPE_ALL` instead?

**Comment by kpet** [comment]
I've just bumped into this as well. I think we should make that change and also allow `CL_DEVICE_TYPE_CUSTOM` devices to be the default device in a platform. The specification currently states:

>The default OpenCL device in the platform. The default OpenCL device must not be a CL_DEVICE_TYPE_CUSTOM device.

which doesn't seem to make sense for a platform that only reports one custom device.

**Comment by bashbaug** [comment]
I'll fix this one after #1069 is merged.
